### ISSUE TITLE
rotate_half RoPE mis-lowers (fp16)

### ISSUE THREAD

**Issue by arielge** [torch-spyre/torch-spyre#3133]
### What happened?

The stock-HuggingFace RoPE form — `x * cos + rotate_half(x) * sin`, where
`rotate_half` **slices along the head dim** (`cat(-x[..., d/2:], x[..., :d/2])`) —
**compiles on Spyre but mis-lowers to wrong numbers** in fp16.

On a single RoPE application (`[B=1, H=16, L=128, head_dim=128]`, fp16, standard
`inv_freq` rotation), the slice form's Spyre output differs from a CPU-fp32
reference by **max_abs_diff = 5.22** (on values of magnitude ~4.3 — i.e. the
output is essentially unrelated to the correct answer).

The identical rotation expressed as a matmul against a `[2, 2, head_dim/2]`
rotation matrix (no head-dim slicing) lowers correctly on the same inputs:
**max_abs_diff = 0.0095**. The two forms are **bit-identical on CPU-fp32**.

**Note on scope vs #912.** #912 ("Rope forward support in Inductor-Spyre",
closed/completed) added RoPE support by hooking FMS `RotaryEmbedding.adjusted_qk`
into a native `spyre.rope_fwd` fusion. That fix does **not** cover an arbitrary
stock-HuggingFace `transformers` forward that reaches `rotate_half` directly.

### What did you expect to happen?

The slice form should lower to match the CPU-fp32 reference within fp16
tolerance (the matmul form achieves ~0.01), or fail to compile — not silently
produce numerically wrong output.

Tolerance used: absolute `max_abs_diff < 5e-2` vs a CPU-fp32 reference.

### How can we reproduce it?

```bash
python3 repro_rope_rotate_half_slice_spyre.py --variant matmul  # control: PASS (~0.0095)
python3 repro_rope_rotate_half_slice_spyre.py --variant slice   # bug:     FAIL (~5.2164)
```

```python
import argparse
import torch

DTYPE = torch.float16
DEVICE = torch.device("spyre")
SEED = 0
BATCH, HEADS, SEQ, HEAD_DIM = 1, 16, 128, 128  # stick-aligned; isolates the slice
HALF = HEAD_DIM // 2


def build_cos_sin():
    pos = torch.arange(SEQ, dtype=torch.float32)[:, None]
    inv_freq = 1.0 / (10000.0 ** (torch.arange(HALF).float() / HALF))
    ang = pos * inv_freq[None, :]
    emb = torch.cat([ang, ang], dim=-1)               # [SEQ, HEAD_DIM]
    return emb.cos()[None].to(DTYPE), emb.sin()[None].to(DTYPE)


def build_rope_matrix(cos, sin):
    c = cos[0, :, :HALF]
    s = sin[0, :, :HALF]
    rot = torch.stack([c, -s, s, c], dim=1).view(SEQ, 2, 2, HALF)
    return rot.contiguous().to(DTYPE)[None]


def rotate_half(x):                                    # stock-HF: slices head dim
    x1 = x[..., : x.shape[-1] // 2]
    x2 = x[..., x.shape[-1] // 2 :]
    return torch.cat((-x2, x1), dim=-1)


def rope_slice(x, cos, sin):                           # stock-HF RoPE
    return x * cos[:, None, :, :] + rotate_half(x) * sin[:, None, :, :]


def apply_rope_matmul(x, rope):                        # matmul form, no slicing
    B, H, L, D = x.shape
    x_ = x.transpose(1, 2).reshape(B, L, H, 2, D // 2)
    sf = rope[:, :, None, :, :, :]
    return sf.mul(x_.unsqueeze(-3)).sum(4, keepdim=True).flatten(3).transpose(1, 2)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--variant", required=True, choices=["matmul", "slice"])
    args = ap.parse_args()

    cos, sin = build_cos_sin()
    rope = build_rope_matrix(cos, sin)
    torch.manual_seed(SEED)
    x = torch.randn(BATCH, HEADS, SEQ, HEAD_DIM, dtype=DTYPE)

    # CONTROL 1: the two forms are algebraically identical on CPU-fp32.
    with torch.no_grad():
        out_slice_cpu = rope_slice(x.float(), cos.float(), sin.float())
        out_matmul_cpu = apply_rope_matmul(x.float(), rope.float())
    equiv = (out_slice_cpu - out_matmul_cpu).abs().max().item()
    print(f"[cpu-fp32] rotate_half vs matmul: max_abs_diff={equiv:.6f} "
          f"({'EQUIVALENT' if equiv < 1e-3 else 'MISMATCH'})")
    ref = out_slice_cpu

    # CONTROL 2 / device under test.
    if args.variant == "slice":
        fn = lambda x_, c_, s_, r_: rope_slice(x_, c_, s_)
    else:
        fn = lambda x_, c_, s_, r_: apply_rope_matmul(x_, r_)
    compiled = torch.compile(fn, dynamic=False)
    with torch.no_grad():
        out = compiled(x.to(DEVICE), cos.to(DEVICE), sin.to(DEVICE),
                       rope.to(DEVICE)).to("cpu").float()

    err = (out - ref).abs().max().item()
    thresh = 5e-2
    print(f"[spyre:{args.variant}] max_abs_diff_vs_cpu={err:.4f} "
          f"finite={torch.isfinite(out).all().item()} "
          f"-> {'PASS' if err < thresh else 'FAIL'} (thresh {thresh})")


if __name__ == "__main__":
    main()
```

**Actual observed output** (torch_spyre `7e45168`, fp16, one process per variant):

```
[cpu-fp32] rotate_half form vs matmul form: max_abs_diff=0.000000 (EQUIVALENT)
[spyre:matmul] max_abs_diff_vs_cpu=0.0095 finite=True -> PASS (thresh 0.05)

[cpu-fp32] rotate_half form vs matmul form: max_abs_diff=0.000000 (EQUIVALENT)
[spyre:slice]  max_abs_diff_vs_cpu=5.2164 finite=True -> FAIL (thresh 0.05)
```

### Any environmental details we need to know?

- torch 2.11.0 ; torch_spyre 0.0.1 @ commit `7e45168f1d56ca1cec4889a3e19b14719dcdd23f` (2026-07-03)
- AIU: PF mode

### Anything else we need to know?

**This blocks Mistral-Small-3.1-24B vision (Pixtral tower).** Confirmed at the
model level with an end-to-end ablation: running the full VLM (adapter-on-Spyre
per-step logits vs stock-CPU reference, teacher-forced) with the vision-tower
RoPE as the matmul form gives 5/5 top-1 token agreement, per-step logit cosine
≥ 0.9994, and the correct caption. Swapping **only** the RoPE to the stock-HF
slice form (everything else identical) drops per-step logit cosine to 0.64 at
decode-3 and produces a **wrong caption** — the e2e cosine floor fails.

- Related: #912 (closed — FMS RoPE compile support; does not cover the stock-HF
  slice path's correctness, see the scope note above), #911 (view propagation
  through stickify).
- hf-adapters context: the adapters avoid this entirely by expressing RoPE as
  `apply_rope_matmul` (a `[2,2,head_dim/2]` matmul, no head-dim slice).

### Relevant log output

```shell
# --variant slice, torch_spyre 7e45168, fp16, fresh process
[cpu-fp32] rotate_half form vs matmul form: max_abs_diff=0.000000 (EQUIVALENT)
[spyre:slice] max_abs_diff_vs_cpu=5.2164 finite=True -> FAIL (thresh 0.05)
```

**Comment by gkumbhat** [comment]
noticed this when working with gemma4 multimodal as well.

Repro:
```python
"""#3133, narrowed: `torch.cat` of two half-width slices of a matmul output is wrong.

Shapes are gemma-4-26B-A4B's vision tower: 2520 patches, hidden 1152, head_dim padded
72 -> 128, so this is the tensor a stock `rotate_half` rope would rotate.

Both halves of the matmul output read back correct on their own; only their
concatenation is uncorrelated. The same `cat` on a freshly transferred tensor of
identical shape and dtype is correct, so the trigger is the source being a matmul
output rather than the shapes.
"""

import torch
import torch.nn.functional as F
import torch_spyre  # noqa: F401

torch.set_grad_enabled(False)
torch.manual_seed(0)

DEVICE = torch.device("spyre:0")
L, HID, HEADS, D = 2520, 1152, 16, 128
HALF = D // 2


def swap(t):  # stock `rotate_half`, without the sign flip
    return torch.cat([t[..., HALF:], t[..., :HALF]], dim=-1)


def cos(a, b):
    return float(F.cosine_similarity(a.flatten().float(), b.flatten().float(), dim=0))


for dtype in (torch.bfloat16, torch.float16):
    h = torch.randn(1, L, HID, dtype=dtype) * 0.5
    lin = torch.nn.Linear(HID, HEADS * D, bias=False).to(dtype)

    # fp32 host reference
    ref = F.linear(h.float(), lin.weight.float()).view(1, L, HEADS, D)

    x = lin.to(DEVICE)(h.to(DEVICE)).view(1, L, HEADS, D)  # matmul output, on device
    fresh = ref.to(dtype).to(DEVICE)  # same shape/dtype, not a matmul output

    print(f"--- {dtype} ---")
    for name, got, want in (
        ("x                  ", x, ref),
        ("x[..., half:]      ", x[..., HALF:], ref[..., HALF:]),
        ("x[..., :half]      ", x[..., :HALF], ref[..., :HALF]),
        ("swap(x)            ", swap(x), swap(ref)),
        ("swap(fresh tensor) ", swap(fresh), swap(ref)),
    ):
        print(f"  {name} cosine={cos(got.to('cpu'), want):+.6f}")
```
